atom_gems/AtomTutorials/Assets/VegetationBending/VegetationBending_Common.azsli

```diff
@@ -0,0 +1,199 @@
+    float m_detailFrequency;
+    float m_detailLeafAmplitude;
+    float m_detailBranchAmplitude;
+
+    // Wind for bending. This would typically come from a wind system instead of a material property, but is here as a proof of concept.
+    float m_windX;
+    float m_windY;
+    float m_bendingStrength;
+    float m_windBendingFrequency;
+
+    Sampler m_sampler
+    {
+        AddressU = Wrap;
+        AddressV = Wrap;
+        MinFilter = Linear;
+        MagFilter = Linear;
+        MipFilter = Linear;
+        MaxAnisotropy = 16;
+    };
+
+    Texture2D m_brdfMap;
+
+    Sampler m_samplerBrdf
+    {
+        AddressU = Clamp;
+        AddressV = Clamp;
+        MinFilter = Linear;
+        MagFilter = Linear;
+        MipFilter = Linear;
+    };
+}
+
+// Callback function for ParallaxMapping.azsli
+DepthResult GetDepth(float2 uv, float2 uv_ddx, float2 uv_ddy)
+{
+    return SampleDepthFromHeightmap(MaterialSrg::m_heightmap, MaterialSrg::m_sampler, uv, uv_ddx, uv_ddy);
+}
+
+
+COMMON_OPTIONS_PARALLAX()
+
+bool ShouldHandleParallax()
+{
+    // Parallax mapping's non uniform uv transformations break screen space subsurface scattering, disable it when subsurface scattering is enabled.
+    return !o_enableSubsurfaceScattering && o_parallax_feature_enabled && o_useHeightmap;
+}
+
+bool ShouldHandleParallaxInDepthShaders()
+{
+    // The depth pass shaders need to calculate parallax when the result could affect the depth buffer, or when
+    // parallax could affect texel clipping.
+    return ShouldHandleParallax() && (o_parallax_enablePixelDepthOffset || o_opacity_mode == OpacityMode::Cutout);
+}
+
+option bool o_color_isBound;
+
+// Derived from "Vegetation Procedural Animation and Shading in Crysis" in GPU Gems 3
+// https://developer.nvidia.com/gpugems/gpugems3/part-iii-rendering/chapter-16-vegetation-procedural-animation-and-shading-crysis
+
+// These are per-object values. They would typically be updated on the CPU to avoid duplicated work,
+// but for simplicity they are calculated here for each vertex as a proof of concept
+float4 SetUpWindBending(float currentTime, float4 worldPosition) 
+{
+    float2 wind = float2(MaterialSrg::m_windX, MaterialSrg::m_windY);
+    float bendingStrength = MaterialSrg::m_bendingStrength;
+    float2 amplitude = float2(wind.x * 0.4 + wind.y * 0.2, wind.y * 0.4 - wind.x * 0.2);
+    float2 frequency = float2(MaterialSrg::m_windBendingFrequency, MaterialSrg::m_windBendingFrequency * 1.125);
+    // Using the world position to modify the phase makes it so different trees near each other are at similar but not equal points in the animation, 
+    // so they appear to be reacting to the same wind but at different times as the wind moves through the vegetation.
+    float2 phase = float2(worldPosition.x * 0.08, worldPosition.y * 0.08);
+
+    float2 addBending = float2(sin(currentTime * frequency.x + phase.x) * amplitude.x, sin(currentTime * frequency.y + phase.y) * amplitude.y);
```

Comment:
I suggest vectorizing expressions as a more concise and "idiomatic" way of authoring shader code. Here for example:

```
float2 bendAmount = sin(currentTime * frequency + phase) * amplitude;

float4 result;
result.xy = bendAmount + wind;
result.z = length(wind);
result.w = 0.3 * length(result.xy);
result.xyz *= bendingStrenth * 0.08;
```

Reply:
Elaborating a bit more, compared to C-interfaces for many functions you might find in `<cmath>` or `<math.h>`, most HLSL built-in functions accept types that have variable component count and base type. You can see this in the type description of the documentation page generally, for example: https://docs.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-sin#type-description. Exploiting the type parameter permissiveness helps a lot of reducing the bookkeeping needed to implement expressions and follow the code later.

Reply:
+1

Reply:
Thanks for the tip, I made your example change & edited some lines.